datafusion/physical-plan/src/sorts/stream.rs

```diff
@@ -88,6 +90,9 @@ pub struct RowCursorStream {
     streams: FusedStreams,
     /// Tracks the memory used by `converter`
     reservation: MemoryReservation,
+    /// Allocated rows for each partition, we keep two to allow for buffering one
+    /// in the consumer of the stream
+    rows: Vec<[Option<Arc<Rows>>; 2]>,
```

Comment:
I think calling this `rows` is somewhat confusing. How about `row_buffer`?

Something that might make this code easier to read could be a wrapper,  the major benefit could be to explain what is going on

```suggestion
    rows: ReusedRows,
```

```rust

/// A pair pf Arc<Rows> that can be reused
struct ReusableRows {
  // inner[stream_idx] holds a two Arcs: 
  // .0 is the currently rows
  // .1 is the currently 
  inner: Vec<[Option<Arc<Rows>>; 2]>,
}

impl ReusedRows {
  // return a Rows for writing, 
  // does not clone if the existing rows can be reused
  fn take_next(&mut self, stream_idx: usize) -> Rows {
        Arc::try_unwrap(self.inner[stream_idx][1].take().unwrap())
            .unwrap_or_else(|_| self.converter.empty_rows(0, 0));
   }
  // save the Rows
  fn save(&mut self, stream_idx, rows: Arc<Rows>) {
    self.rows[stream_idx][1] = Some(Arc::clone(&rows));
     let [a, b] = &mut self.rows[stream_idx];
     std::mem::swap(a, b);
  }

```